frontend/src/component/common/NewConstraintAccordion/ConstraintAccordionEdit/ConstraintAccordionEdit.tsx

```diff
@@ -98,63 +100,125 @@ export const ConstraintAccordionEdit = ({
     const [expanded, setExpanded] = useState(false);
     const [action, setAction] = useState('');
 
+    const { input, validator, setError, error } = useConstraintInput({
+        contextDefinition,
+        localConstraint,
+    });
+
     useEffect(() => {
         // Setting expanded to true on mount will cause the accordion
         // animation to take effect and transition the expanded accordion in
         setExpanded(true);
     }, []);
 
-    useEffect(() => {
-        if (onAutoSave) {
-            onAutoSave(localConstraint);
-        }
-    }, [JSON.stringify(localConstraint)]);
-
     useEffect(() => {
         setContextDefinition(
             resolveContextDefinition(context, localConstraint.contextName),
         );
     }, [localConstraint.contextName, context]);
 
+    useEffect(() => {
+        setError('');
+    }, [setError]);
+
+    const onUndo = () => {
+        if (constraintChanges.length < 2) return;
+        const previousChange = constraintChanges[constraintChanges.length - 2];
+
+        setLocalConstraint(previousChange);
+        setConstraintChanges((prev) => prev.slice(0, prev.length - 1));
+    };
+
+    const recordChange = (localConstraint: IConstraint) => {
+        setConstraintChanges((prev) => [...prev, localConstraint]);
+
+        if (
+            onAutoSave &&
+            localConstraint.values &&
+            localConstraint.values.length > 0
+        ) {
+            onAutoSave(localConstraint);
+        }
+
+        if (onAutoSave && localConstraint.value) {
+            onAutoSave(localConstraint);
+        }
+    };
+
     const setContextName = useCallback((contextName: string) => {
-        setLocalConstraint((prev) => ({
-            ...prev,
-            contextName,
-            values: [],
-            value: '',
-        }));
+        setLocalConstraint((prev) => {
+            const localConstraint = cleanConstraint({
+                ...prev,
+                contextName,
+                values: [],
+                value: '',
+            });
+
+            recordChange(localConstraint);
+
+            return localConstraint;
+        });
     }, []);
 
     const setOperator = useCallback((operator: Operator) => {
-        setLocalConstraint((prev) => ({
-            ...prev,
-            operator,
-            values: [],
-            value: '',
-        }));
+        setLocalConstraint((prev) => {
+            const localConstraint = cleanConstraint({
+                ...prev,
+                operator,
+                values: [],
+                value: '',
+            });
+
+            recordChange(localConstraint);
+
+            return localConstraint;
+        });
     }, []);
 
-    const setValues = useCallback((values: string[]) => {
+    const setValues = useCallback((values: string[], record = true) => {
```

Comment:
maybe another function setValuesWithRecord instead of setValues with boolean flag?